### PR TITLE
[Reputation Oracle] Add auth and login

### PR DESCRIPTION
**Pull request by flopez7**
## Description

Add authentication in reputation oracle

## How test the changes

`yarn test`

## Operational checklist

- [x] All new functionality is covered by tests
- [ ] Any related documentation has been changed or added

### CODE REVIEW

File: packages/apps/reputation-oracle/server/src/common/config/env.ts [hunk 1]
```diff
@@ -52,6 +61,10 @@ export const envValidator = Joi.object({
   S3_SECRET_KEY: Joi.string().required(),
   S3_BUCKET: Joi.string().default('launcher'),
   S3_USE_SSL: Joi.string().default(false),
+  // SendGrid
+  SENDGRID_API_KEY: Joi.string().required(),
+  SENDGRID_FROM_EMAIL: Joi.string().default('job-launcher@hmt.ai'),
+  SENDGRID_FROM_NAME: Joi.string().default('Human Protocol Job Launcher'),
```
Comment (leric7) [inline packages/apps/reputation-oracle/server/src/common/config/env.ts:67]:
```suggestion
  SENDGRID_FROM_EMAIL: Joi.string().default('reputation-oracle@hmt.ai'),
  SENDGRID_FROM_NAME: Joi.string().default('Human Protocol Reputation Oracle'),
```

File: packages/apps/reputation-oracle/server/src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      user: userEntity,
+    });
+
+    await this.sendgridService.sendEmail({
+      to: data.email,
+      subject: 'Verify your email',
+      html: `Welcome to the Job Launcher Service.<br />
+Click <a href="${this.feURL}/verify?token=${tokenEntity.uuid}">here</a> to complete sign up.`,
+      text: `Welcome to the Job Launcher Service.
+Click ${this.feURL}/verify?token=${tokenEntity.uuid} to complete sign up.`,
```
Comment (leric7) [inline packages/apps/reputation-oracle/server/src/modules/auth/auth.service.ts:91]:
```suggestion
      html: `Welcome to the Reputation Oracle Service.<br />
Click <a href="${this.feURL}/verify?token=${tokenEntity.uuid}">here</a> to complete sign up.`,
      text: `Welcome to the Reputation Oracle Service.
Click ${this.feURL}/verify?token=${tokenEntity.uuid} to complete sign up.`,
```

File: packages/apps/reputation-oracle/server/src/modules/auth/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+      user: userEntity,
+    });
+
+    this.sendgridService.sendEmail({
+      to: data.email,
+      subject: 'Verify your email',
+      html: `Welcome to the Job Launcher Service.<br />
+Click <a href="${this.feURL}/verify?token=${tokenEntity.uuid}">here</a> to complete sign up.`,
+      text: `Welcome to the Job Launcher Service.
+Click ${this.feURL}/verify?token=${tokenEntity.uuid} to complete sign up.`,
```
Comment (leric7) [inline packages/apps/reputation-oracle/server/src/modules/auth/auth.service.ts:215]:
```suggestion
      html: `Welcome to the Reputation Oracle Service.<br />
Click <a href="${this.feURL}/verify?token=${tokenEntity.uuid}">here</a> to complete sign up.`,
      text: `Welcome to the Reputation Oracle Service.
Click ${this.feURL}/verify?token=${tokenEntity.uuid} to complete sign up.`,
```

File: packages/apps/reputation-oracle/server/src/modules/webhook/webhook.cron.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export class WebhookCron {
     private readonly webhookRepository: WebhookRepository,
   ) {}
 
-  @Cron(CronExpression.EVERY_10_SECONDS)
+  // @Cron(CronExpression.EVERY_10_SECONDS)
```
Comment (leric7) [inline packages/apps/reputation-oracle/server/src/modules/webhook/webhook.cron.ts:19]: Remove commented line, or add TODO comment if we need this later.

File: packages/apps/reputation-oracle/server/.env.example [hunk 5]
```diff
@@ -4,12 +4,13 @@ HOST=localhost
 PORT=3008
 FE_URL=http://localhost:3009
 SESSION_SECRET=test
+PASSWORD_SECRET=
```
Comment (portuu3) [inline packages/apps/reputation-oracle/server/.env.example:7]: I think it is not necessary

File: packages/apps/reputation-oracle/server/src/common/enums/user.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export enum UserStatus {
+  ACTIVE = 'ACTIVE',
+  INACTIVE = 'INACTIVE',
+  PENDING = 'PENDING',
+}
+
+export enum UserType {
```
Comment (portuu3) [inline packages/apps/reputation-oracle/server/src/common/enums/user.ts:7]: For this case userTypes would be Operator, Exchange Oracle, Recording Oracle and Worker

File: packages/apps/reputation-oracle/server/src/modules/webhook/webhook.controller.ts [hunk 7]
```diff
@@ -1,10 +1,12 @@
...
+import { Body, Controller, Post, Request, UseGuards } from '@nestjs/common';
+import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
 import { Public } from '../../common/decorators';
 import { WebhookService } from './webhook.service';
 import { WebhookIncomingDto } from './webhook.dto';
+import { JwtAuthGuard } from 'src/common/guards/jwt.auth';
 
-@Public()
+@ApiBearerAuth()
+@UseGuards(JwtAuthGuard)
```
Comment (portuu3) [inline packages/apps/reputation-oracle/server/src/modules/webhook/webhook.controller.ts:9]: this endpoint doesn't need to be protected with bearer token

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/apps/reputation-oracle/server/src/modules/auth/auth.controller.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+@ApiTags('Auth')
+@Controller('/auth')
+export class AuthJwtController {
+  constructor(private readonly authService: AuthService) {}
+
+  @Public()
+  @Post('/signup')
+  @UseInterceptors(ClassSerializerInterceptor)
+  public async signup(@Body() data: UserCreateDto): Promise<void> {
```
Comment (portuu3) [inline packages/apps/reputation-oracle/server/src/modules/auth/auth.controller.ts:35]: UserCreateDto cannot include the status, in the sign up, the user could set any status here.

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/apps/reputation-oracle/server/src/modules/auth/auth.controller.ts (hunk 8 above)
Comment (flopez7) [reply packages/apps/reputation-oracle/server/src/modules/auth/auth.controller.ts:35]: How should we do it? Assigning Operator by default and then the user type needs to be updated?

**Review by portuu3 — APPROVED** [review]

**Pull request merged**